### PR TITLE
Closes #2738 Include 3rd Party CSS into concatenation - exclude_css sanitization (3/3) 

### LINKED ISSUE #2738: Include 3rd Party CSS into concatenation - exclude_css sanitization (3/3)
Sub-task for epic #2689 

### `exclude_css` sanitization
Edit the sanitization of exclude_css input field and allow external URL.

Effort S

### PR DESCRIPTION
**Pull request by crystinutzaa**
Sub-task for epic #2689 

### `exclude_css` sanitization
Edit the sanitization of exclude_css input field and allow external URL.

Effort S

### CODE REVIEW

File: inc/functions/formatting.php [hunk 1]
```diff
@@ -102,6 +102,25 @@ function rocket_validate_js( $file ) {
...
+ */
+function rocket_validate_css( $file ) {
+	if ( rocket_is_internal_file( $file ) ) {
+		$file = trim( $file );
+		$file = rocket_clean_exclude_file( $file );
+
+		return $file;
+	}
+
+	return sanitize_text_field( \rocket_remove_url_protocol( strtok( $file, '?' ) ) );
```
Comment (Tabrisrp) [inline inc/functions/formatting.php:121]:
```suggestion
	return sanitize_text_field( rocket_remove_url_protocol( strtok( $file, '?' ) ) );
```

File: tests/Unit/inc/Engine/Admin/Settings/Settings/sanitizeCallback.php [hunk 2]
```diff
@@ -101,6 +101,33 @@ public function testShouldSanitizeFontPreloadEntries( $input, $expected ) {
 		);
 	}
 
+	/**
+	 * @dataProvider addExcludeCSSProvider
+	 */
+	public function testShouldSanitizeExcludeCSS( $original, $sanitized ) {
+		Functions\when( 'wp_parse_url' )->alias( function( $url, $component ) {
+			return parse_url( $url, $component );
+		} );
```
Comment (hellofromtonya) [inline tests/Unit/inc/Engine/Admin/Settings/Settings/sanitizeCallback.php:110]: For this use the `StubTrait:: stubWpParseUrl()` instead.

File: inc/functions/formatting.php [hunk 3]
```diff
@@ -102,6 +102,25 @@ function rocket_validate_js( $file ) {
...
+ *
+ * @param  string $file filepath to sanitize.
+ * @return string
+ */
+function rocket_validate_css( $file ) {
+	if ( rocket_is_internal_file( $file ) ) {
+		$file = trim( $file );
+		$file = rocket_clean_exclude_file( $file );
+
+		return $file;
```
Comment (hellofromtonya) [inline inc/functions/formatting.php:118]: Let's combine these to be more concise and performant.

```suggestion
		return rocket_clean_exclude_file( trim( $file ) );
```

**Review by hellofromtonya — CHANGES_REQUESTED** [review]
Need unit tests for this new function `rocket_validate_css()`.

**Comment by crystinutzaa** [comment]
@hellofromtonya do we really need Unit test for that new function?
The current Unit test & Integration which I added calls this function and we can consider that is tested.

**Comment by hellofromtonya** [comment]
@crystinutzaa Yes we do. Why? Unit tests test in isolation. The test for the sanitizer callback should mock this new function and any other functions too.

File: tests/Unit/inc/Engine/Admin/Settings/Settings/sanitizeCallback.php [hunk 4]
```diff
@@ -101,6 +100,31 @@ public function testShouldSanitizeFontPreloadEntries( $input, $expected ) {
...
+	 * @dataProvider addExcludeCSSProvider
+	 */
+	public function testShouldSanitizeExcludeCSS( $original, $sanitized ) {
+		$this->stubWpParseUrl();
+
+		Functions\when( 'content_url' )->justReturn( 'http://example.org/wp-content/' );
+		Functions\when( 'get_rocket_i18n_uri' )->justReturn( [ 'http://example.org/' ] );
+
+		Functions\when( 'sanitize_text_field' )->returnArg();
+		Functions\when( 'rocket_valid_key' )->justReturn( true );
```
Comment (hellofromtonya) [inline tests/Unit/inc/Engine/Admin/Settings/Settings/sanitizeCallback.php:113]: @crystinutzaa You'll want to also mock the new `rocket_validate_css()` function too as it's a dependency to this callback but external to it.

**Review by hellofromtonya — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Review by hellofromtonya — APPROVED** [review]

**Comment by crystinutzaa** [comment]
@wp-media/dev-qa ready for testing

**Review by arunbasillal — APPROVED** [review]
- Tried adding `https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css` with master branch to Exclude CSS text area. Result 
image

- Repeated the same with feature branch. Result 
image

@crystinutzaa That is the only test needed for this PR right? It is not supposed to do anything in the front end or anywhere else I believe

**Comment by crystinutzaa** [comment]
@arunbasillal yes, this is the only test needed 👍

**Pull request merged**